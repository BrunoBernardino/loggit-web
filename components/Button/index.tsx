import React, { forwardRef } from 'react';
import styled from 'styled-components';

import './style.scss';

interface ButtonProps {
  element?: 'button' | 'a';
  isDisabled?: boolean;
  type?: 'primary' | 'secondary' | 'delete';
  href?: string;
  onClick?: any;
  className?: string;
  width?: '' | 'large' | 'tiny';
}

const Button: React.FC<ButtonProps> = forwardRef(
  (props: ButtonProps, ref: any) => {
    const {
      element,
      className,
      type,
      isDisabled,
      width,
      ...remainingProps
    } = props;

    if (element === 'button') {
      const StyledButton = styled.button.attrs({
        className: `${className || ''} Button Button--${type} ${
          width ? `Button--${width}` : ''
        }`,
      })``;
      return (
        <StyledButton
          ref={ref}
          type="button"
          disabled={isDisabled}
          {...remainingProps}
        />
      );
    }

    const StyledAnchor = styled.a.attrs({
      className: `${className || ''} Button Button--${type} ${
        width ? `Button--${width}` : ''
      }`,
    })``;

    return <StyledAnchor ref={ref} {...remainingProps} />;
  },
);

Button.defaultProps = {
  element: 'button',
  isDisabled: false,
  type: 'primary',
  width: '',
};

export default Button;
