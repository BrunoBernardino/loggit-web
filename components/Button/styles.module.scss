@import 'styles/__variables';

$transition-speed: 140ms;

a.Button {
  display: inline-block;
  padding: 1.3rem 1.8rem;

  &--large {
    padding: 1.3rem 4.2rem;
  }

  &--tiny {
    font-size: 0.85rem;
    line-height: 1.2rem;
    margin: 0 auto;
    text-align: center;
    padding: 0.5rem 0.8rem;
    font-weight: normal;
  }
}

button.Button {
  display: block;
  padding: 1rem 1.8rem;

  &--large {
    padding: 1rem 4.2rem;
  }

  &--tiny {
    padding: 1rem;
  }
}

.Button {
  color: $color-white;
  font-size: 1.3rem;
  line-height: 1.5rem;
  margin: 0 auto;
  text-align: center;
  padding: 1rem 1.3rem;
  border-radius: 5px;
  cursor: pointer;
  text-decoration: none;
  font-weight: 500;
  z-index: 0;
  position: relative;
  background-size: 100%;
  transition: all $transition-speed ease-in;
  border: none;

  &:focus,
  &:hover {
    color: $color-white;
    box-shadow: 1px 1px 3px $color-block-border;
    text-shadow: 1px 1px 1px $color-text;
    outline: none;

    opacity: 0.7;
  }

  &--primary {
    background-color: $color-menu-background;
  }

  &--secondary {
    background-color: $color-light-gray;

    @media (prefers-color-scheme: dark) {
      background-color: $color-text-gray;
    }
    @at-root .theme-dark #{&} {
      background-color: $color-text-gray;
    }

    &:focus,
    &:hover {
      box-shadow: 1px 1px 3px $color-block-border;
      text-shadow: 1px 1px 1px $color-block-border;
    }
  }

  &--delete {
    background-color: $color-background-red;
  }
}
