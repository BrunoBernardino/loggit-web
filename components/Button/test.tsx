import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import expect from 'expect';
import enzymify from 'expect-enzyme';

import Button from './index';

expect.extend(enzymify());

describe('Button', () => {
  it('renders the button without errors', () => {
    const wrapper = shallow(<Button>Hello</Button>);
    expect(wrapper.first().text()).toBe('Hello');
  });

  it('renders the button as expected', () => {
    const tree = renderer.create(<Button>Hello</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
