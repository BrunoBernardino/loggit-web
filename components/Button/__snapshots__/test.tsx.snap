// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders the button as expected 1`] = `
<button
  className="Button__StyledButton-kxixwi-0  Button Button--primary "
  disabled={false}
  type="button"
>
  Hello
</button>
`;
